asserts/snapasserts/snapasserts.go

```diff
@@ -113,6 +133,110 @@ func CrossCheck(instanceName, snapSHA3_384, provenance string, snapSize uint64,
 	return snapRev, nil
 }
 
+// CrossCheckResource tries to cross check the name, hash digest, size,
+// provenance, and metadata of a snap resource with the relevant assertions
+// (snap-resource-revision and snap-resource-pair) in a database that should be
+// pre-populated with them.
+func CrossCheckResource(name, hash, provenance string, size uint64, csi *snap.ComponentSideInfo, si *snap.SideInfo, model *asserts.Model, db Finder) error {
+	headers := map[string]string{
+		"resource-sha3-384": hash,
+		"resource-name":     name,
+		"snap-id":           si.SnapID,
+	}
+	if provenance != "" {
+		headers["provenance"] = provenance
+	}
+
+	a, err := db.Find(asserts.SnapResourceRevisionType, headers)
+	if err != nil {
+		provInf := ""
+		if provenance != "" {
+			provInf = fmt.Sprintf(" provenance: %s", provenance)
+		}
+		return fmt.Errorf("internal error: cannot find pre-populated snap-resource-revision assertion for %q: %s%s", name, hash, provInf)
+	}
+
+	resrev := a.(*asserts.SnapResourceRevision)
+
+	if resrev.ResourceSize() != size {
+		return fmt.Errorf(
+			"resource %q file does not have expected size according to signatures (download is broken or tampered): %d != %d",
+			name, size, resrev.ResourceSize(),
+		)
+	}
+
+	if resrev.SnapID() != si.SnapID || resrev.ResourceRevision() != csi.Revision.N {
```

Comment:
Is the ID part of this check necessary given that it's part of the primary key?

Reply:
no indeed, in this snap-resource-revision are different from snaps as we have more than just an hash in their primary key